e2e/xbridge/token.go

```diff
@@ -0,0 +1,181 @@
+package xbridge
+
+import (
+	"context"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/e2e/app/eoa"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+	"github.com/omni-network/omni/lib/netconf"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+type tokenDeploymentConfig struct {
+	Config     deploymentConfig
+	Name       string
+	Symbol     string
+	Minter     common.Address
+	Admin      common.Address
+	Upgrader   common.Address
+	Pauser     common.Address
+	Clawbacker common.Address
+}
+
+func (cfg tokenDeploymentConfig) validateTokenConfig() error {
+	if err := cfg.Config.validateDeploymentConfig(); err != nil {
+		return errors.Wrap(err, "validate config")
+	}
+	if cfg.Name == "" {
+		return errors.New("name is empty")
+	}
+	if cfg.Symbol == "" {
+		return errors.New("symbol is empty")
+	}
+	if isEmpty(cfg.Minter) {
+		return errors.New("minter is zero")
+	}
+	if isEmpty(cfg.Admin) {
+		return errors.New("admin is zero")
+	}
+	if isEmpty(cfg.Upgrader) {
+		return errors.New("upgrader is zero")
+	}
+	if isEmpty(cfg.Pauser) {
+		return errors.New("pauser is zero")
+	}
+	if isEmpty(cfg.Clawbacker) {
+		return errors.New("clawbacker is zero")
+	}
+
+	return nil
+}
+
+// tokenAddress returns the token contract address for the given network.
+func tokenAddress(ctx context.Context, network netconf.ID, deployment tokenDescriptors) (common.Address, error) {
+	return contracts.Create3Address(ctx, network, deployment.symbol)
+}
+
+// tokenSalt returns the salt for the token contract for the given network.
+func tokenSalt(ctx context.Context, network netconf.ID, deployment tokenDescriptors) (string, error) {
+	return contracts.Create3Salt(ctx, network, deployment.symbol)
+}
+
+// deployTokenIfNeeded deploys a new rlusd contract if it is not already deployed.
+// If the contract is already deployed, the receipt is nil.
+func DeployTokenIfNeeded(ctx context.Context, network netconf.ID, backend *ethbackend.Backend, wrapper bool, deployment xBridgeDeployment) (common.Address, *ethtypes.Receipt, error) {
+	targetDeployment := deployment.token
+	if wrapper {
+		targetDeployment = deployment.wrapped
+	}
```

Comment:
Instead of passing a flag `wrapped` to modify it's behavior. Consider passing in target deployment explicitly

```golang
DeployIfNeededI(ctx, network, backend, deployment.token)
// or
DeployIfNeededI(ctx, network, backend, deployment.wrapped)
```

I see these args are passed down a number of times so might not be as easy as that.